src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Caching.cs

```diff
@@ -23,10 +23,55 @@ public sealed partial class JsonSerializerOptions
         // Simple LRU cache for the public (de)serialize entry points that avoid some lookups in _cachingContext.
         private volatile JsonTypeInfo? _lastTypeInfo;
 
+        /// <summary>
+        /// Gets the <see cref="JsonTypeInfo"/> contract metadata resolved by the current <see cref="JsonSerializerOptions"/> instance.
+        /// </summary>
+        /// <param name="type">The type to resolve contract metadata for.</param>
+        /// <returns>The contract metadata resolved for <paramref name="type"/>.</returns>
+        /// <remarks>
+        /// Returned metadata can be downcast to <see cref="JsonTypeInfo{T}"/> and used with the relevant <see cref="JsonSerializer"/> overloads.
+        ///
+        /// If the <see cref="JsonSerializerOptions"/> instance is locked for modification, the method will return a cached instance for the metadata.
+        /// </remarks>
+        [RequiresUnreferencedCode("Getting a metadata for a type may require reflection which depends on unreferenced code.")]
+        [RequiresDynamicCode("Getting a metadata for a type may require reflection which depends on runtime code generation.")]
+        public JsonTypeInfo GetTypeInfo(Type type)
+        {
+            if (type is null)
+            {
+                ThrowHelper.ThrowArgumentNullException(nameof(type));
+            }
+
+            if (JsonTypeInfo.IsInvalidForSerialization(type))
+            {
+                ThrowHelper.ThrowArgumentException_CannotSerializeInvalidType(nameof(type), type, null, null);
+            }
+
+            _typeInfoResolver ??= DefaultJsonTypeInfoResolver.RootDefaultInstance();
+
+            JsonTypeInfo? typeInfo;
+            if (IsLockedInstance)
+            {
+                typeInfo = GetCachingContext()?.GetOrAddJsonTypeInfo(type);
+                typeInfo?.EnsureConfigured();
+            }
+            else
+            {
+                typeInfo = GetTypeInfoNoCaching(type);
+            }
+
+            if (typeInfo is null)
+            {
+                ThrowHelper.ThrowNotSupportedException_NoMetadataForType(type);
+            }
```

Comment:
I'm wondering if we should return null instead, that would match IJsonTypeInfoResolver. Also I just noticed that our docs for GetConverter claim we return null when there is no converter but they also claim we throw - we might want to match behavior and fix docs:
https://docs.microsoft.com/en-us/dotnet/api/system.text.json.jsonserializeroptions.getconverter?view=net-6.0

Reply:
I think we should make the docs match with the shipped behavior.

Reply:
either way is fine but we should be consistent at least

Reply:
Sure, but this is not related to the scope of this change in the slightest way possible.